lib/surrounded/context/name_collision_detector.rb

```diff
@@ -0,0 +1,43 @@
+module NameCollisionDetector
+
+
+  def on_name_collision(handler)
+    case handler
+    when :nothing
+      @handler = lambda {|role, array| }
+    when :raise_exception
+      @handler = lambda {|role, array| raise Surrounded::Context::NameCollisionError.new "#{role} has name collisions with #{array}"}
+    when :warn
+      @handler = lambda {|role, array| puts "#{role} has name collisions with #{array}" if array.length > 0}
+    else
+      @handler = handler
+    end
```

Comment:
using a case statement can lock us into a particular set of possibilities.
if you alter the design of this slightly, you can allow any method that handles the same method signature.
I was thinking about something like this:
```ruby
def on_name_collision(handler)
  if handler.respond_to?(:call)
    @name_collision_handler = handler
  else
    @name_collision_handler = instance_method(handler)
  end
end
```
We could then expect that the handler would accept a single argument of the exception

```ruby
class SomeContext
  on_name_collision :raise # call built-in Kernel.raise
  on_name_collision :warn # call built-in Kernel.warn
  on_name_collision :do_something # call a custom method
  on_name_collision ->(message){ STDERR.puts(message) } # use a Proc

  def do_something(message)
    # custom implementation here
  end
end
```

The only downside with that, I think, is that `warn` accepts a string and `raise` can accept an exception or a string or both... So perhaps a case statement will be in order.
Have any thoughts on that?